include/TPP/IR/StructuredOpMatcher.h

```diff
@@ -115,6 +115,26 @@ struct HasStaticShape {
   }
 };
 
+// Callable object to verify `operand` to have a rank in `ranks`.
+struct HasRank {
+  HasRank() = delete;
+  explicit HasRank(std::initializer_list<int64_t> ranks) : ranks(ranks){};
+
+  bool operator()(OpOperand *operand, Operation *op) const {
+    auto operandType = operand->get().getType();
+    if (!isa<ShapedType>(operandType))
+      return llvm::is_contained(ranks, HasRank::SCALAR);
+    int64_t rank = operandType.cast<ShapedType>().getRank();
+    return llvm::any_of(
+        ranks, [=](int64_t expectedRank) { return expectedRank == rank; });
+  }
+
+  // There are multiple way to represent a scalar: f32, tensor<f32>.
+  // SCALAR means f32.
+  static constexpr int64_t SCALAR = -1;
```

Comment:
What does `-1` means here?

Reply:
represents a scalar (i.e. f32 or non shaped type).

Reply:
Well, that's the obvious answer. My point is, why `-1`? Why not `0`?

Why have the special constant when the others are just integers with the number of dims?

Why not just construct this with `{ 0, 1, 2 }` (which is perfectly obvious to me) instead of `{HasRank::SCALAR, 1, 2}`?

Reply:
Thinking a bit more here. It is better to keep them separate until we support memref<f32> or tensor<f32> otherwise it is unclear what { 0 } means. Should we match f32 or memref<f32>/tensor<f32>?

Reply:
It is not all types, just shapedType vs scalarType. -1 represents a scalar type. Both input are representative: `f32` and `memref<f32>`.

Reply:
What is the input you're working on?

Reply:
The input (test case in `linalg-to-tpp-memref/tensor.mlir` with `memref<f32>` was provided by Adam when moving the default pipeline to tensor. The fact that we need to handle `f32` is because we do have tests that require to handle scalars (https://github.com/plaidml/tpp-mlir/blob/5c19269f0fcdaddb009a9337de258fa5457bbb8d/test/Integration/copy.mlir#L60).

Reply:
That's an artificial IR from a test, I'm not sure what's the relevance of it. We need to understand what is the meaning of `memref<f32>`, ie. what to front-ends mean when they lower to this shape and now `f32`.

Reply:
Deciding to support `f32` and/or `memref<f32>` can be addresses in another PR. This PR intends not to allow `memref<f32>` as tpp verifier requires `f32` to specify a scalar and not `memref<f32>`. The change is needed when moving the default pipeline to tpp on tensor.

Reply:
The solution may be to add a check for the particular case of `memref<f32>` instead of forcing all TPP ops to output 2D shapes or to create an artificial rule about all of their inputs.

It'd be much simpler to just bail on `memref<f32>` and leave a comment like `TODO: check if this is a common pattern and what to do about it`. If this isn't coming from any ingress IR, then it's not high priority to fix this now.